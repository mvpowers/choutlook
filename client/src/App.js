import React, { Component } from 'react';
import styled from 'styled-components';
import openSocket from 'socket.io-client';
import { Helmet } from 'react-helmet';
import {
  Navbar,
  NewMessageBar,
  FolderList,
  IconsBar,
  MessageList,
  MessageView,
} from './components';

const crypto = require('crypto');
const uuidv4 = require('uuid/v4');

const serverHost = process.env.SERVER_HOST || '127.0.0.1';
const serverPort = process.env.SERVER_PORT || '3000';
const socket = openSocket(`http://${serverHost}:${serverPort}`);

console.warn('Socket connect to:', `http://${serverHost}:${serverPort}`);

const Container = styled.div`
  display: flex;
  flex-direction: column;
  flex: 1;
  height: 100%;
`;

const Content = styled.div`
  display: flex;
  flex-direction: row;
  flex: 1;
  height: 100%;
  overflow: hidden;
`;

export default class App extends Component {
  constructor() {
    super();
    this.state = {
      displayMsg: [],
      sendMsg: '',
      password: '',
      username: 'anonymous',
      editingUsername: false,
      replyFocus: false,
      unreadMsg: false,
      sessionId: uuidv4(),
    };
  }

  componentDidMount() {
    const { password, replyFocus, sessionId } = this.state;
    socket.on('broadcastMessage', data => {
      const decipher = crypto.createDecipher('aes192', password);
      let decrypted = decipher.update(data, 'hex', 'utf8');
      try {
        decrypted += decipher.final('utf8');
        decrypted = JSON.parse(decrypted);
      } catch (e) {
        decrypted = { username: 'admin', sendMsg: 'Error decrypting data' };
      }
      this.setState({
        displayMsg: [
          ...this.state.displayMsg,
          {
            user: decrypted.username,
            message: decrypted.sendMsg,
            sessionId: decrypted.sessionId,
            time: Date.now().toString(),
          },
        ],
      });
      if (!replyFocus && decrypted.sessionId !== sessionId) {
        this.setState({ unreadMsg: true });
      }
    });

    socket.on('userConnect', () => {
      this.setState({
        displayMsg: [
          ...this.state.displayMsg,
          {
            user: 'admin',
            message: 'User has entered chat',
            time: Date.now().toString(),
          },
        ],
      });
    });

    socket.on('userDisconnect', () => {
      this.setState({
        displayMsg: [
          ...this.state.displayMsg,
          {
            user: 'admin',
            message: 'User has left chat',
            time: Date.now().toString(),
          },
        ],
      });
    });
  }

  toggleEditUsername = () => {
    this.setState(prevState => ({
      editingUsername: !prevState.editingUsername,
    }));
  };

  handleChange = e => {
    this.setState({ [e.target.id]: e.target.value });
  };

  handleSend = () => {
    const { password, sendMsg, username, sessionId } = this.state;
    const msgObj = JSON.stringify({ username, sendMsg, sessionId });
    const cipher = crypto.createCipher('aes192', password);
    let encrypted = cipher.update(msgObj, 'utf8', 'hex');
    encrypted += cipher.final('hex');

    socket.emit('createMessage', encrypted);
    this.setState({ sendMsg: '' });
  };

  handleDiscard = () => {
    this.setState({ sendMsg: '' });
  };

  handleKeyPress = e => {
    const { password, sendMsg, username, sessionId } = this.state;

    this.setState({ unreadMsg: false });

    if (!e.shiftKey && e.key === 'Enter') {
      const msgObj = JSON.stringify({ username, sendMsg, sessionId });
      const cipher = crypto.createCipher('aes192', password);
      let encrypted = cipher.update(msgObj, 'utf8', 'hex');
      encrypted += cipher.final('hex');

      socket.emit('createMessage', encrypted);
      this.setState({ sendMsg: '' });
    }
  };

  handleFocus = () => {
    this.setState({ replyFocus: true, unreadMsg: false });
  };

  handleBlur = () => {
    this.setState({ replyFocus: false, unreadMsg: false });
  };

  render() {
    const {
      displayMsg,
      sendMsg,
      password,
      username,
      replyFocus,
      editingUsername,
      unreadMsg,
    } = this.state;
    return (
      <Container>
        <Navbar password={password} handleChange={this.handleChange} />
        <NewMessageBar />
        <Content>
          <IconsBar />
          <FolderList />
          <MessageList displayMsg={displayMsg} replyFocus={replyFocus} />
          <MessageView
            sendMsg={sendMsg}
            username={username}
            editingUsername={editingUsername}
            toggleEditUsername={this.toggleEditUsername}
            handleChange={this.handleChange}
            submitSendMsg={this.handleSend}
            discardSendMsg={this.handleDiscard}
            submitOnEnter={this.handleKeyPress}
            handleFocus={this.handleFocus}
            handleBlur={this.handleBlur}
          />
        </Content>
        <Helmet>
          <title>{unreadMsg ? 'Choutlook*' : 'Choutlook'}</title>
        </Helmet>
      </Container>
    );
  }
}
