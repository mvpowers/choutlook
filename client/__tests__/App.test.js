import React from 'react';
import Enzyme, { shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import App from '../src/App';

Enzyme.configure({ adapter: new Adapter() });

describe('App', () => {
  it('should be defined', () => {
    expect(App).toBeDefined();
  });
  it('should render correctly', () => {
    const tree = shallow(<App />);
    expect(tree).toMatchSnapshot();
  });
});
