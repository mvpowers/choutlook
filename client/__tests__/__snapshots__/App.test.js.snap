// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render correctly 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "class",
    "props": Object {
      "children": Array [
        <Navbar
          handleChange={[Function]}
          password=""
        />,
        <Unknown />,
        <App__Content>
          <Unknown />
          <Unknown />
          <MessageList
            displayMsg={Array []}
            replyFocus={false}
          />
          <MessageView
            discardSendMsg={[Function]}
            editingUsername={false}
            handleBlur={[Function]}
            handleChange={[Function]}
            handleFocus={[Function]}
            sendMsg=""
            submitOnEnter={[Function]}
            submitSendMsg={[Function]}
            toggleEditUsername={[Function]}
            username="anonymous"
          />
        </App__Content>,
        <HelmetWrapper
          defer={true}
          encodeSpecialCharacters={true}
        >
          <title>
            Choutlook
          </title>
        </HelmetWrapper>,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "handleChange": [Function],
          "password": "",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": Array [
            <Unknown />,
            <Unknown />,
            <MessageList
              displayMsg={Array []}
              replyFocus={false}
            />,
            <MessageView
              discardSendMsg={[Function]}
              editingUsername={false}
              handleBlur={[Function]}
              handleChange={[Function]}
              handleFocus={[Function]}
              sendMsg=""
              submitOnEnter={[Function]}
              submitSendMsg={[Function]}
              toggleEditUsername={[Function]}
              username="anonymous"
            />,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "class",
            "props": Object {
              "displayMsg": Array [],
              "replyFocus": false,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "discardSendMsg": [Function],
              "editingUsername": false,
              "handleBlur": [Function],
              "handleChange": [Function],
              "handleFocus": [Function],
              "sendMsg": "",
              "submitOnEnter": [Function],
              "submitSendMsg": [Function],
              "toggleEditUsername": [Function],
              "username": "anonymous",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
        ],
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "children": <title>
            Choutlook
          </title>,
          "defer": true,
          "encodeSpecialCharacters": true,
        },
        "ref": null,
        "rendered": Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Choutlook",
          },
          "ref": null,
          "rendered": "Choutlook",
          "type": "title",
        },
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "class",
      "props": Object {
        "children": Array [
          <Navbar
            handleChange={[Function]}
            password=""
          />,
          <Unknown />,
          <App__Content>
            <Unknown />
            <Unknown />
            <MessageList
              displayMsg={Array []}
              replyFocus={false}
            />
            <MessageView
              discardSendMsg={[Function]}
              editingUsername={false}
              handleBlur={[Function]}
              handleChange={[Function]}
              handleFocus={[Function]}
              sendMsg=""
              submitOnEnter={[Function]}
              submitSendMsg={[Function]}
              toggleEditUsername={[Function]}
              username="anonymous"
            />
          </App__Content>,
          <HelmetWrapper
            defer={true}
            encodeSpecialCharacters={true}
          >
            <title>
              Choutlook
            </title>
          </HelmetWrapper>,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "handleChange": [Function],
            "password": "",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": Array [
              <Unknown />,
              <Unknown />,
              <MessageList
                displayMsg={Array []}
                replyFocus={false}
              />,
              <MessageView
                discardSendMsg={[Function]}
                editingUsername={false}
                handleBlur={[Function]}
                handleChange={[Function]}
                handleFocus={[Function]}
                sendMsg=""
                submitOnEnter={[Function]}
                submitSendMsg={[Function]}
                toggleEditUsername={[Function]}
                username="anonymous"
              />,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "class",
              "props": Object {
                "displayMsg": Array [],
                "replyFocus": false,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "discardSendMsg": [Function],
                "editingUsername": false,
                "handleBlur": [Function],
                "handleChange": [Function],
                "handleFocus": [Function],
                "sendMsg": "",
                "submitOnEnter": [Function],
                "submitSendMsg": [Function],
                "toggleEditUsername": [Function],
                "username": "anonymous",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
          ],
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "children": <title>
              Choutlook
            </title>,
            "defer": true,
            "encodeSpecialCharacters": true,
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Choutlook",
            },
            "ref": null,
            "rendered": "Choutlook",
            "type": "title",
          },
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
